src/dynamic_contract.rs

```diff
@@ -0,0 +1,280 @@
+const PLATFORM_FILE_PREFIX: &str = "lib";
+/// Dynamic link library prefix
+#[cfg(windows)]
+const PLATFORM_FILE_PREFIX: &str = "";
+/// Dynamic link library file extension specific to the platform
+#[cfg(any(target_os = "macos", target_os = "ios"))]
+const PLATFORM_FILE_EXTENSION: &str = "dylib";
+/// Dynamic link library file extension specific to the platform
+#[cfg(all(unix, not(any(target_os = "macos", target_os = "ios"))))]
+const PLATFORM_FILE_EXTENSION: &str = "so";
+/// Dynamic link library file extension specific to the platform
+#[cfg(windows)]
+const PLATFORM_FILE_EXTENSION: &str = "dll";
+
+/// Creates a platform-specific file path
+fn create_library_path(name: &str) -> PathBuf {
+    let mut path = PathBuf::new();
+    path.push("target");
+    path.push(CARGO_PROFILE);
+    path.push("deps");
+    path.push(PLATFORM_FILE_PREFIX.to_string() + name);
+    path.set_extension(PLATFORM_FILE_EXTENSION);
+    path
+}
+
+// All contracts export a symbol named process()
+const ENTRYPOINT: &str = "process";
+
+type Entrypoint = unsafe extern "C" fn(tx: &Transaction, accounts: &mut [Account]) -> Result<()>;
+
+pub const DYNAMIC_CONTRACT_ID: [u8; 32] = [
+    2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+];
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub enum DynamicContract {
+    /// Native contract
+    /// * Transaction::keys[0..] - contract dependent
+    /// * name - name of the contract, translated to a file path of the contract module
+    /// * userdata - contract specific user data
+    Native { name: String, userdata: Vec<u8> },
+    /// BPF contract
+    /// * Transaction::keys[0..] - contract dependent
+    /// * TODO BPF specific stuff
+    /// * userdata - contract specific user data
+    BPF { userdata: Vec<u8> },
+}
+
+impl DynamicContract {
+    pub fn id() -> Pubkey {
+        Pubkey::new(&DYNAMIC_CONTRACT_ID)
+    }
+
+    pub fn check_id(contract_id: &Pubkey) -> bool {
+        contract_id.as_ref() == DYNAMIC_CONTRACT_ID
+    }
+
+    pub fn get_balance(account: &Account) -> i64 {
+        account.tokens
+    }
+
+    pub fn process_transaction(tx: &Transaction, accounts: &mut [Account]) {
+        // TODO how should contract code/data be conveyed (currently userdata)
+        // TODO loading the contract each time we want to call it sucks
+        let dc: DynamicContract = deserialize(&tx.userdata).unwrap();
+        match dc {
+            DynamicContract::Native { name, .. } => {
+                let path = create_library_path(&name);
+                let lib = Library::new(&path).expect("Failed to load library");
+                unsafe {
+                    let entrypoint: Symbol<Entrypoint> = match lib.get(ENTRYPOINT.as_bytes()) {
+                        Ok(s) => s,
+                        Err(e) => panic!("{:?} Unable to find {:?} in {:?}", e, ENTRYPOINT, &path),
+                    };
+                    if let Err(e) = entrypoint(tx, accounts) {
```

Comment:
The contract only needs `tx.userdata` and `accounts`, no need to pass in the entire `Transaction`